lib/grandpa/message_handler.go

```diff
@@ -280,21 +303,50 @@ func (h *MessageHandler) verifyCommitMessageJustification(fm *CommitMessage) err
 }
 
 func (h *MessageHandler) verifyPreVoteJustification(msg *CatchUpResponse) (common.Hash, error) {
+	voters := make(map[ed25519.PublicKeyBytes]map[common.Hash]int, len(msg.PreVoteJustification))
+	eqVotesByHash := make(map[common.Hash]map[ed25519.PublicKeyBytes]struct{})
+
+	// identify equivocatory votes by hash
+	for _, justification := range msg.PreVoteJustification {
+		if _, ok := voters[justification.AuthorityID]; !ok {
+			voters[justification.AuthorityID] = make(map[common.Hash]int)
+		}
+
+		voters[justification.AuthorityID][justification.Vote.Hash]++
+
+		if voters[justification.AuthorityID][justification.Vote.Hash] > 1 {
+			if _, ok := eqVotesByHash[justification.Vote.Hash]; !ok {
+				eqVotesByHash[justification.Vote.Hash] = make(map[ed25519.PublicKeyBytes]struct{})
+			}
+
+			eqVotesByHash[justification.Vote.Hash][justification.AuthorityID] = struct{}{}
+		}
```

Comment:
nit just splitting steps since this is a bit hardcore to read

```suggestion
		hashToCount, ok := voters[justification.AuthorityID]
		if !ok {
			hashToCount = make(map[common.Hash]int)
		}

		hashToCount[justification.Vote.Hash]++

		voters[justification.AuthorityID] = hashToCount

		if hashToCount[justification.Vote.Hash] > 1 {
			publicKeysSet, ok := eqVotesByHash[justification.Vote.Hash]
			if !ok {
				publicKeysSet = make(map[ed25519.PublicKeyBytes]struct{})
			}

			publicKeysSet[justification.AuthorityID] = struct{}{}
			eqVotesByHash[justification.Vote.Hash] = publicKeysSet
		}
```

Reply:
also would it be possible to use `getEquivocatoryVoters` here instead?